hub_test.go

```diff
@@ -3596,59 +3596,159 @@ func TestJoinRoomSwitchClient(t *testing.T) {
 }
 
 func TestGetRealUserIP(t *testing.T) {
-	REMOTE_ATTR := "192.168.1.2"
-
-	trustedProxies, err := ParseAllowedIps("192.168.0.0/16")
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	request := &http.Request{
-		RemoteAddr: REMOTE_ATTR + ":23456",
-	}
-	if ip := GetRealUserIP(request, trustedProxies); ip != REMOTE_ATTR {
-		t.Errorf("Expected %s but got %s", REMOTE_ATTR, ip)
-	}
-
-	X_REAL_IP := "192.168.10.11"
-	request.Header = http.Header{
-		http.CanonicalHeaderKey("x-real-ip"): []string{X_REAL_IP},
-	}
-	if ip := GetRealUserIP(request, trustedProxies); ip != X_REAL_IP {
-		t.Errorf("Expected %s but got %s", X_REAL_IP, ip)
-	}
-
-	// "X-Real-IP" has preference before "X-Forwarded-For"
-	X_FORWARDED_FOR_IP := "192.168.20.21"
-	X_FORWARDED_FOR := X_FORWARDED_FOR_IP + ":12345, 192.168.30.32"
-	request.Header = http.Header{
-		http.CanonicalHeaderKey("x-real-ip"):       []string{X_REAL_IP},
-		http.CanonicalHeaderKey("x-forwarded-for"): []string{X_FORWARDED_FOR},
-	}
-	if ip := GetRealUserIP(request, trustedProxies); ip != X_REAL_IP {
-		t.Errorf("Expected %s but got %s", X_REAL_IP, ip)
-	}
-
-	request.Header = http.Header{
-		http.CanonicalHeaderKey("x-forwarded-for"): []string{X_FORWARDED_FOR},
-	}
-	if ip := GetRealUserIP(request, trustedProxies); ip != X_FORWARDED_FOR_IP {
-		t.Errorf("Expected %s but got %s", X_FORWARDED_FOR_IP, ip)
+	testcases := []struct {
+		expected string
+		headers  http.Header
+		trusted  string
+		addr     string
+	}{
+		{
+			"192.168.1.2",
+			nil,
+			"192.168.0.0/16",
+			"192.168.1.2:23456",
+		},
+		{
+			"10.11.12.13",
+			nil,
+			"192.168.0.0/16",
+			"10.11.12.13:23456",
+		},
+		{
+			"10.11.12.13",
+			http.Header{
+				http.CanonicalHeaderKey("x-real-ip"): []string{"10.11.12.13"},
+			},
+			"192.168.0.0/16",
+			"192.168.1.2:23456",
+		},
+		{
+			"11.12.13.14",
+			http.Header{
+				http.CanonicalHeaderKey("x-forwarded-for"): []string{"11.12.13.14, 192.168.30.32"},
+			},
+			"192.168.0.0/16",
+			"192.168.1.2:23456",
+		},
+		// "X-Real-IP" has preference before "X-Forwarded-For"
+		{
+			"10.11.12.13",
+			http.Header{
+				http.CanonicalHeaderKey("x-real-ip"):       []string{"10.11.12.13"},
+				http.CanonicalHeaderKey("x-forwarded-for"): []string{"11.12.13.14, 192.168.30.32"},
+			},
+			"192.168.0.0/16",
+			"192.168.1.2:23456",
+		},
+		// Multiple "X-Forwarded-For" headers are merged.
+		{
+			"11.12.13.14",
+			http.Header{
+				http.CanonicalHeaderKey("x-forwarded-for"): []string{"11.12.13.14", "192.168.30.32"},
```

Comment:
I think it would be better if three or more headers were tested and the returned IP was in one of the middle headers, as right now the code could just use the first header without actually joining all of them and the test would still pass (of course the current code joins all of them and is fine, but just in case it is changed at some point :-) ).

Reply:
Added in #750